File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let ans = [];
    let id = 0;
    let left = 0;
    let right = matrix[0].length - 1;
    let up = 0;
    let down = matrix.length - 1;

    while (up <= down && left <= right) {
        if (id === 0) {
            for (let i = left; i <= right; i++) {
                ans.push(matrix[up][i]);
            }
            up++;
        } else if (id === 1) {
            for (let i = up; i <= down; i++) {
                ans.push(matrix[i][right]);
            }
            right--;
        } else if (id === 2) {
            for (let i = right; i >= left; i--) {
                ans.push(matrix[down][i]);
            }
            down--;
        } else {
            for (let i = down; i >= up; i--) {
                ans.push(matrix[i][left]);
            }
            left++;
        }
        id = (id + 1) % 4;
    }

    return ans;
};

File: 0054-spiral-matrix/0054-spiral-matrix.rs
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let mut ans = Vec::new();
        let mut id = 0;
        let mut left: i32 = 0;
        let mut right: i32 = (matrix[0].len() as i32) - 1;
        let mut up: i32 = 0;
        let mut down: i32 = (matrix.len() as i32) - 1;

        while up <= down && left <= right {
            if id == 0 {
                for i in left..=right {
                    ans.push(matrix[up as usize][i as usize]);
                }
                up += 1;
            } else if id == 1 {
                for i in up..=down {
                    ans.push(matrix[i as usize][right as usize]);
                }
                right -= 1;
            } else if id == 2 {
                for i in (left..=right).rev() {
                    ans.push(matrix[down as usize][i as usize]);
                }
                down -= 1;
            } else {
                for i in (up..=down).rev() {
                    ans.push(matrix[i as usize][left as usize]);
                }
                left += 1;
            }
            id = (id + 1) % 4;
        }

        ans
    }
}
